iroh-bytes/src/store/file.rs

```diff
@@ -0,0 +1,2483 @@
+//!
+//! Errors:
+//!
+//! ActorError is an enum containing errors that can happen inside message
+//! handlers of the actor. This includes various redb related errors and io
+//! errors when reading or writing non-inlined data or outboard files.
+//!
+//! OuterError is an enum containing all the actor errors and in addition
+//! errors when communicating with the actor.
+use std::{
+    collections::{btree_map, BTreeMap, BTreeSet},
+    io::{self, BufReader, Read},
+    path::{Path, PathBuf},
+    sync::{Arc, RwLock},
+    time::SystemTime,
+};
+
+use bao_tree::io::{
+    fsm::Outboard,
+    outboard::PostOrderMemOutboard,
+    sync::{ReadAt, Size},
+};
+use bytes::Bytes;
+use futures::{channel::oneshot, Stream, StreamExt};
+
+use iroh_base::hash::{BlobFormat, Hash, HashAndFormat};
+use iroh_io::AsyncSliceReader;
+use redb::{AccessGuard, ReadableTable, StorageError};
+use serde::{Deserialize, Serialize};
+use smallvec::SmallVec;
+use tokio::io::AsyncWriteExt;
+use tracing::trace_span;
+
+mod import_flat_store;
+mod tables;
+#[cfg(test)]
+mod tests;
+mod util;
+mod validate;
+
+use crate::{
+    store::{
+        bao_file::{BaoFileStorage, CompleteMemOrFileStorage},
+        file::util::{overwrite_and_sync, read_and_remove, ProgressReader},
+    },
+    util::{
+        progress::{IdGenerator, IgnoreProgressSender, ProgressSendError, ProgressSender},
+        LivenessTracker, MemOrFile,
+    },
+    Tag, TempTag, IROH_BLOCK_SIZE,
+};
+use tables::{
+    ReadOnlyTables, ReadableTables, Tables, BLOBS_TABLE, INLINE_DATA_TABLE, INLINE_OUTBOARD_TABLE,
+};
+
+use super::{
+    bao_file::{raw_outboard_size, BaoFileConfig, BaoFileHandle, CreateCb, DropCb},
+    temp_name, BaoBatchWriter, EntryStatus, ExportMode, ExportProgressCb, ImportMode,
+    ImportProgress, MapEntry, ReadableStore, TempCounterMap, ValidateProgress,
+};
+
+use super::BaoBlobSize;
+
+/// Location of the data.
+///
+/// Data can be inlined in the database, a file conceptually owned by the store,
+/// or a number of external files conceptually owned by the user.
+///
+/// Only complete data can be inlined.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub(crate) enum DataLocation<I = (), E = ()> {
+    /// Data is in the inline_data table.
+    Inline(I),
+    /// Data is in the canonical location in the data directory.
+    Owned(E),
+    /// Data is in several external locations. This should be a non-empty list.
+    External(Vec<PathBuf>, E),
+}
+
+impl<X> DataLocation<X, u64> {
+    fn union(self, that: DataLocation<X, u64>) -> io::Result<Self> {
+        Ok(match (self, that) {
+            (DataLocation::External(mut paths, size), DataLocation::External(b_paths, ..)) => {
+                paths.extend(b_paths);
+                paths.sort();
+                paths.dedup();
+                DataLocation::External(paths, size)
+            }
+            (DataLocation::External(_, _), b @ DataLocation::Owned(_)) => b,
+            (a, _b) => a,
```

Comment:
this seems inconsistent, if I union `Internal` with `External` I get `Internal` but if I call it the otherway around I get `External`

Reply:
why?

external/external -> 
external/owned -> owned (b)
owned/owned -> owned (a)
owned/external -> owned (a)

anyway, spelled it out a bit clearer now.

Reply:
```
(external, internal) => external
(internal, external) => internal
```